lib/paperclip/storage/s3v2.rb

```diff
@@ -0,0 +1,466 @@
+    #   link in the +url+ entry for more information about S3 domains and buckets.
+    # * +url+: There are four options for the S3 url. You can choose to have the bucket's name
+    #   placed domain-style (bucket.s3.amazonaws.com) or path-style (s3.amazonaws.com/bucket).
+    #   You can also specify a CNAME (which requires the CNAME to be specified as
+    #   :s3_alias_url. You can read more about CNAMEs and S3 at
+    #   http://docs.amazonwebservices.com/AmazonS3/latest/index.html?VirtualHosting.html
+    #   Normally, this won't matter in the slightest and you can leave the default (which is
+    #   path-style, or :s3_path_url). But in some cases paths don't work and you need to use
+    #   the domain-style (:s3_domain_url). Anything else here will be treated like path-style.
+    #
+    #   Notes:
+    #   * The value of this option is a string, not a symbol.
+    #     <b>right:</b> <tt>":s3_domain_url"</tt>
+    #     <b>wrong:</b> <tt>:s3_domain_url</tt>
+    #   * If you use a CNAME for use with CloudFront, you can NOT specify https as your
+    #     :s3_protocol;
+    #     This is *not supported* by S3/CloudFront. Finally, when using the host
+    #     alias, the :bucket parameter is ignored, as the hostname is used as the bucket name
+    #     by S3. The fourth option for the S3 url is :asset_host, which uses Rails' built-in
+    #     asset_host settings.
+    #   * To get the full url from a paperclip'd object, use the
+    #     image_path helper; this is what image_tag uses to generate the url for an img tag.
+    # * +path+: This is the key under the bucket in which the file will be stored. The
+    #   URL will be constructed from the bucket and the path. This is what you will want
+    #   to interpolate. Keys should be unique, like filenames, and despite the fact that
+    #   S3 (strictly speaking) does not support directories, you can still use a / to
+    #   separate parts of your file name.
+    # * +s3_region+: The region.
+    # * +s3_metadata+: These key/value pairs will be stored with the
+    #   object.  This option works by prefixing each key with
+    #   "x-amz-meta-" before sending it as a header on the object
+    #   upload request. Can be defined both globally and within a style-specific hash.
+    # * +s3_storage_class+: If this option is set to
+    #   <tt>:reduced_redundancy</tt>, the object will be stored using Reduced
+    #   Redundancy Storage.  RRS enables customers to reduce their
+    #   costs by storing non-critical, reproducible data at lower
+    #   levels of redundancy than Amazon S3's standard storage.
+    #
+    #   You can set storage class on a per style bases by doing the following:
+    #     :s3_storage_class => {
+    #       :thumb => :reduced_reduncancy
+    #     }
+    #   Or globally:
+    #     :s3_storage_class => :reduced_redundancy
+
+    module S3v2
+      def self.extended base
```

Comment:
Assignment Branch Condition size for extended is too high. [98.38/15]<br>Use def with parentheses when there are parameters.